inference-engine/thirdparty/clDNN/src/gpu/strided_slice_gpu.cpp

```diff
@@ -77,10 +77,14 @@ struct strided_slice_gpu : typed_primitive_gpu_impl<strided_slice> {
         // If the ith bit of end_mask is not set, end[i] is ignored and the fullest possible range in that dimension is used
         // instead.
         vector_assign_if_not_mask(params.striding_params[1], out_shape, params.end_mask);
-
         for (size_t dim = 0; dim < params.striding_params[2].size(); dim++) {
-            auto begin = params.striding_params[0][dim] < 0 ? out_shape[dim] + params.striding_params[0][dim] : params.striding_params[0][dim];
-            auto end = params.striding_params[1][dim] < 0 ? out_shape[dim] + params.striding_params[1][dim] : params.striding_params[1][dim];
+            if (params.striding_params[0][dim] < 0)
+                params.striding_params[0][dim] = out_shape[dim] + params.striding_params[0][dim];
+            if (params.striding_params[1][dim] < 0)
+                params.striding_params[1][dim] = out_shape[dim] + params.striding_params[1][dim];
+
+            auto begin = params.striding_params[0][dim];
+            auto end = params.striding_params[1][dim];
             auto stride = params.striding_params[2][dim];
             if (stride < 0 && (end > begin)) {
                 std::swap(params.striding_params[0][dim], params.striding_params[1][dim]);
```

Comment:
I'm not sure why we have to support cases where stride is negative and yet end > begin, but I think reducing "new begin" (the one that we got after swapping begin ↔ end) by 1 is not always right. While it probably should happen when "new begin" >= out_shape[dim], for other cases swapping is IMHO enough, like for X dimension here:

-  Input (BFYX):  1x3x4x6
- Begin (BFYX):  0x2x0x1
- End (BFYX):    1x3x2x4
- Stride (BFYX): 1x1x1x-2
- Output (BFYX): 1x1x2x2

Reply:
As I understand this branch is needed only when reversed order is needed, begin_mask=1 && end_mask=1, and the values in `begin` and `end` are ignored, so begin=0; end=out_shape[dim]; Looks like for this case this code works correctly (added comment with the sample + some test more cases)

At least the following tests case:
```
        stridedSliceParamsTuple({ 1, 3, 4, 6 }, { 0, 2, 0, 1 }, { 1, 3, 2, 4 }, { 1, 1, 1, -2 },
                                { 0, 0, 0, 0 }, { 0, 0, 0, 0 },  { },  { },  { },
                                InferenceEngine::Precision::FP32, CommonTestUtils::DEVICE_GPU),
```
fails with `C++ exception with description "StridedSlice_22278 has zero dimension that is not allowable" thrown in the test body.` exception from ngraph

Reply:
I agree that this part doesn't look crucial (looks that such reversing not supported in numpy or tensorflow anyway) and don't want it to be a blocker here, but I still think this reversed order is not covered properly:
- right now, only 'begin' and 'end' local variables are being swapped and they are not used later - we should rather operate on params.striding_params[0][dim] and params.striding_params[1][dim] instead,
- I don't see a reason to changing these values if they are both equal in the first place.

Generated numbers for following additional unit test cases are different than the one that I get from numpy/TF (assuming that reversing the order of indices was expected behavior, note that we won't be able to select element on 0 index anyway, as putting -1 in 'end' has different meaning):
- input=np.arange(0,7,1), fwd=input[0:7:3]={0,3,6}             <-> bwd=input[6:0:-3]={6,3} 
- input=np.arange(0,7,1), fwd=input[0:8:3]={0,3,6}             <-> bwd=input[6:0:-3]={6,3}
- input=np.arange(0,7,1), fwd=input[1:7:3]={1,4}               <-> bwd=input[4:0:-3]={4,1}
- input=np.arange(0,7,1), fwd=input[1:8:3]={1,4,7}             <-> bwd=input[7:0:-3]={7,4,1}
- input=np.arange(0,7,1), fwd=input[2:7:3]={2,5}               <-> bwd=input[5:1:-3]={5,2}
- input=np.arange(0,7,1), fwd=input[2:8:3]={2,5}               <-> bwd=input[5:1:-3]={5,2}
- input=np.arange(0,7,1), fwd=input[3:7:3]={3,6}               <-> bwd=input[6:2:-3]={6,3}
- input=np.arange(0,7,1), fwd=input[3:8:3]={3,6}               <-> bwd=input[6:2:-3]={6,3}
- input=np.arange(0,7,1), fwd=input[0:7:1]={0,1,2,3,4,5,6}     <-> bwd=input[6:0:-1]={6,5,4,3,2,1}
- input=np.arange(0,7,1), fwd=input[0:8:1]={0,1,2,3,4,5,6,7}   <-> bwd=input[7:0:-1]={7,6,5,4,3,2,1}
- input=np.arange(0,7,1), fwd=input[0:7:2]={0,2,4,6}           <-> bwd=input[6:0:-2]={6,4,2}
- input=np.arange(0,7,1), fwd=input[0:8:2]={0,2,4,6}           <-> bwd=input[6:0:-2]={6,4,2}
- input=np.arange(0,7,1), fwd=input[1:7:2]={1,3,5}             <-> bwd=input[5:0:-2]={5,3,1}
- input=np.arange(0,7,1), fwd=input[1:8:2]={1,3,5,7}           <-> bwd=input[7:0:-2]={7,5,3,1}
- input=np.arange(0,7,1), fwd=input[2:7:2]={2,4,6}             <-> bwd=input[6:1:-2]={6,4,2}
- input=np.arange(0,7,1), fwd=input[2:8:2]={2,4,6}             <-> bwd=input[6:1:-2]={6,4,2}

Following code was used to calculate reversed begin/end values:
```
if ((stride < 0) && (begin < end)) {
    cldnn::tensor::value_type new_begin = std::min(end, out_shape[dim] - 1);
    cldnn::tensor::value_type new_end = begin;
    if (((end - begin) % (-stride)) == 0) {
        new_begin = std::max(end - 1, 0);
    }
    new_begin = begin + ((new_begin - new_end) / (-stride)) * (-stride);
    new_end = std::max(new_end - 1, 0);
    params.striding_params[0][dim] = new_begin;
    params.striding_params[1][dim] = new_end;
}
```

Reply:
@jhajducz 
>right now, only 'begin' and 'end' local variables are being swapped and they are not used later - we should rather operate on params.striding_params[0][dim] and params.striding_params[1][dim] instead

the type is `auto&`, so it's a ref to the values in `params.striding_params[0][dim]` and swap/decrement is supposed to change them as well.

>I don't see a reason to change these values if they are both equal in the first place.

Yeah, looks like if begin==end, then we have 0 output shape

> Following code was used to calculate reversed begin/end values

Well, lets compare the output of the current version and the code provided by you:
```
[ RUN      ] smoke_CLDNN/StridedSliceLayerTest.CompareWithRefs/inShape=(1.12.100)_netPRC=FP32_begin=(0.9.0)_end=(0.4.0)_stride=(-1.-2.-1)_begin_m=(1.0.1)_end_m=(1.0.1)_new_axis_m=(0.0.0)_shrink_m=(0.0.0)_ellipsis_m=(0.0.0)_targetDevice=GPU

AXIS:0 SHAPE: 100
INITIAL: 0; 1
NEW: 0; 0
OLD: 0; -1
AXIS:2
INITIAL: 0; 100
NEW: 99; 0
OLD: 99; -1
```
INITIAL - is what we have before any manipulations, NEW - your code, OLD - code in this PR, pair of values - begin and end respectively.
For axis 2 we have begin=0; end=100; stride=-1 which means that we need to put slices [99;0] in reversed order to the output; Current logic returns 99; -1 which is pretty much what is expected (note: `-1 != out_shape[dim]-1` anymore, it's basically the range of the "loop" to get required data, because negative begin/end values are processed a bit earlier). New logic returns 99; 0, so the test passes, but `end` value correctness is questionable (if we include it to the work range - it's correct).

Now consider this case:
```
[ RUN      ] smoke_CLDNN/StridedSliceLayerTest.CompareWithRefs/inShape=(2.2.4.2)_netPRC=FP32_begin=(1.0.0.0)_end=(1.2.4.2)_stride=(1.1.-2.-1)_begin_m=(0.1.1.1)_end_m=(1.1.1.1)_new_axis_m=def_shrink_m=def_ellipsis_m=def_targetDevice=GPU

AXIS:2 SHAPE: 4
INITIAL: 0; 4
NEW: 2; 0
OLD: 3; -1
AXIS:3 SHAPE: 2
INITIAL: 0; 2
NEW: 1; 0
OLD: 1; -1
```
Here we have difference in `end` values for both axes again, but you can also see that `begin` values for axis=2 are also different. So with your code we don't include 3rd slice to the output at all which leads to the test fail.

Reply:
> the type is `auto&`, so it's a ref to the values in `params.striding_params[0][dim]` and swap/decrement is supposed to change them as well.

Yeah, now it's ok, in initial version type was just "auto", I missed the fact that you've already fix it, sorry.



> note: `-1 != out_shape[dim]-1` anymore, it's basically the range of the "loop" to get required data, because negative begin/end values are processed a bit earlier

If that special case is handled before, then I guess setting end to -1 is okay. Then 3rd line from end in my code should be changed to just decrement "new_end".

After such fix in my code, I still can see some suspicious test cases. For example, there are differences in elements that are chosen in a following set https://github.com/jhajducz/openvino/commit/db5c444c39f9d24a55f66ae39085a420406881b9. All of them assume unit dimenions except for last (X) axis, where we have values input=np.arange(0,7,1). I'll list the indices that would have been chosen for this axis if corresponding stride was positive as 'fwd', indices that are chosen by your code as 'bwd_before' and indices that are chosen by my code as 'bwd'. The actual element values given in {} brackets were calculated with numpy/TF (except for the cases where end=-1, as its still being treated as out_shape[dim]-1 there, so I've calculated them manually). Even though all of these test cases are failing for both code variants, at least 'bwd' indices look more correct to me than 'bwd_before' (unless I still miss the general goal here and reflecting fwd indices for positive stride is not a desired behavior):

- strided_slice_gpu_f32_i32.test_1x1x1x8_negative_stride_begin_smaller_than_end2
```
fwd=input[0:8:3]={0,3,6}
bwd=input[6:-1:-3]={6,3,0}
bwd_before=input[7:-1:-3]={7,4,1}
```

- strided_slice_gpu_f32_i32.test_1x1x1x8_negative_stride_begin_smaller_than_end3
```
fwd=input[1:7:3]={1,4}
bwd=input[4:0:-3]={4,1}
bwd_before=input[6:0:-3]={6,3}
```

- strided_slice_gpu_f32_i32.test_1x1x1x8_negative_stride_begin_smaller_than_end5
```
fwd=input[2:7:3]={2,5}
bwd=input[5:1:-3]={5,2}
bwd_before=input[6:1:-3]={6,3}

```
- strided_slice_gpu_f32_i32.test_1x1x1x8_negative_stride_begin_smaller_than_end6
```
fwd=input[2:8:3]={2,5}
bwd=input[5:1:-3]={5,2}
bwd_before=input[7:1:-3]={7,4}
```

- strided_slice_gpu_f32_i32.test_1x1x1x8_negative_stride_begin_smaller_than_end8
```
fwd=input[3:8:3]={3,6}
bwd=input[6:2:-3]={6,3}
bwd_before=input[7:2:-3]={7,4}
```

- strided_slice_gpu_f32_i32.test_1x1x1x8_negative_stride_begin_smaller_than_end12
```
input[0:8:2]={0,2,4,6}
bwd=input[6:-1:-2]={6,4,2,0}
bwd_before=input[7:-1:-2]={7,5,3,1}
```

- strided_slice_gpu_f32_i32.test_1x1x1x8_negative_stride_begin_smaller_than_end13
```
fwd=input[1:7:2]={1,3,5}
bwd=input[5:0:-2]={5,3,1}
bwd_before=input[6:0:-2]={6,3]
```

- strided_slice_gpu_f32_i32.test_1x1x1x8_negative_stride_begin_smaller_than_end16
```
fwd=input[2:8:2]={2,4,6}
bwd=input[6:1:-2]={6,4,2}
bwd_before=input[7:1:-2]={7,5,3}

```

Reply:
> unless I still miss the general goal here and reflecting fwd indices for positive stride is not a desired behavior

As I can see, expected output is not just reversed fwd pass, and bwd_before from your example matches with the reference from ngraph and cpu plugin impl
```
fwd=input[2:8:2]={2,4,6}
bwd_before=input[7:1:-2]={7,5,3}
```

So looks like negative stride and begin=0 means that we have to start slicing from out_shape[dim] - 1 element and then move in reversed order with given stride.

Reply:
I see, then I got it wrong in the first place, thanks for explanation.